pkg/api/agentPoolOnlyApi/v20180331/validate.go

```diff
@@ -141,108 +141,121 @@ func validateVNET(a *Properties) error {
+			return e
 		}
+	} else {
+		return ErrorNilAgentPoolProfile
+	}
 
-		if a.AgentPoolProfiles == nil {
-			return ErrorNilAgentPoolProfile
-		}
+	return nil
+}
 
-		// validate custom VNET logic
-		if isCustomVNET(a.AgentPoolProfiles) {
-			var subscription string
-			var resourceGroup string
-			var vnet string
+func validateAgentPoolVNET(a []*AgentPoolProfile) error {
 
-			for _, agentPool := range a.AgentPoolProfiles {
-				// validate subscription, resource group and vnet are the same among subnets
-				subnetSubscription, subnetResourceGroup, subnetVnet, _, err := GetVNETSubnetIDComponents(agentPool.VnetSubnetID)
-				if err != nil {
-					return ErrorParsingSubnetID
-				}
+	// validate custom VNET logic at agent pool level
+	if isCustomVNET(a) {
+		var subscription string
+		var resourceGroup string
+		var vnet string
 
-				if subscription == "" {
-					subscription = subnetSubscription
-				} else {
-					if subscription != subnetSubscription {
-						return ErrorSubscriptionNotMatch
-					}
-				}
+		for _, agentPool := range a {
+			// validate each agent pool has a subnet
+			if !agentPool.IsCustomVNET() {
+				return ErrorAtLeastAgentPoolNoSubnet
+			}
 
-				if resourceGroup == "" {
-					resourceGroup = subnetResourceGroup
-				} else {
-					if resourceGroup != subnetResourceGroup {
-						return ErrorResourceGroupNotMatch
-					}
+			// validate subscription, resource group and vnet are the same among subnets
+			subnetSubscription, subnetResourceGroup, subnetVnet, _, err := GetVNETSubnetIDComponents(agentPool.VnetSubnetID)
+			if err != nil {
+				return ErrorParsingSubnetID
+			}
+
+			if subscription == "" {
+				subscription = subnetSubscription
+			} else {
+				if subscription != subnetSubscription {
+					return ErrorSubscriptionNotMatch
 				}
+			}
 
-				if vnet == "" {
-					vnet = subnetVnet
-				} else {
-					if vnet != subnetVnet {
-						return ErrorVnetNotMatch
-					}
+			if resourceGroup == "" {
+				resourceGroup = subnetResourceGroup
+			} else {
+				if resourceGroup != subnetResourceGroup {
+					return ErrorResourceGroupNotMatch
 				}
 			}
-		} else {
-			return ErrorAgentPoolNoSubnet
-		}
 
-	} else if string(n.NetworkPlugin) == string(Kubenet) {
-		if n.ServiceCidr != "" || n.DNSServiceIP != "" || n.DockerBridgeCidr != "" {
-			return ErrorKubenetNoCustomization
+			if vnet == "" {
+				vnet = subnetVnet
+			} else {
+				if vnet != subnetVnet {
+					return ErrorVnetNotMatch
+				}
+			}
 		}
-	} else {
-		return ErrorInvalidNetworkPlugin
 	}
 
 	return nil
 }
 
+// check agent pool subnet, return true as long as one agent pool has a subnet defined.
 func isCustomVNET(a []*AgentPoolProfile) bool {
-
 	for _, agentPool := range a {
-		if !agentPool.IsCustomVNET() {
-			return false
+		if agentPool.IsCustomVNET() {
+			return true
 		}
 	}
 
-	return true
+	return false
 }
 
 // GetVNETSubnetIDComponents extract subscription, resourcegroup, vnetname, subnetname from the vnetSubnetID
```

Comment:
The regex is static. Can be saved at package level. Also, most literals used in Azure are case insensitive. For instance, `resourcegroups` is as good as `resourceGroups`

Reply:
Yes, we will look into this in a separate task.